### ISSUE TITLE
Implement Real-Time Delay & Arrival Time Predictions

### ISSUE THREAD

**Issue by trackrat-dev** [trackrat-dev/TrackRat#154]
## Overview

Implement a comprehensive delay and arrival time prediction system for TrackRat that leverages historical journey data to provide accurate, real-time predictions for train arrivals. This system will help riders make informed decisions about their journeys by predicting delays and providing confidence intervals for arrival times.

## Background

TrackRat currently collects extensive journey data including:
- Scheduled vs actual departure/arrival times
- Journey stops with timing information
- Historical delay patterns across routes
- Real-time train position data

This rich dataset can be leveraged to build predictive models that forecast delays and arrival times based on historical patterns and current conditions.

### 1. Core Prediction Service

Create a new service in `backend_v2/src/trackrat/services/predictions.py`:

```python
class ArrivalPredictionService:
    async def predict_arrival(
        self,
        journey_id: int,
        target_station: str,
        current_station: Optional[str] = None,
        current_delay: Optional[int] = None
    ) -> ArrivalPrediction:
        # Core prediction logic
        pass
```

### 2. Database Schema Enhancements

Add new tables to track predictions and features:

```sql
-- Table for storing prediction results and accuracy tracking
CREATE TABLE prediction_results (
    id INTEGER PRIMARY KEY,
    journey_id INTEGER REFERENCES train_journeys(id),
    prediction_type TEXT DEFAULT 'arrival',
    target_station_code TEXT,
    predicted_at TIMESTAMP,
    prediction_data JSON, -- {predicted_arrival, confidence_interval, delay_minutes}
    actual_outcome JSON, -- {actual_arrival, actual_delay}
    confidence_score FLOAT,
    model_version TEXT,
    created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
);

-- Table for feature engineering
CREATE TABLE journey_features (
    journey_id INTEGER PRIMARY KEY REFERENCES train_journeys(id),
    hour_of_day INTEGER,
    day_of_week INTEGER,
    is_rush_hour BOOLEAN,
    is_holiday BOOLEAN,
    is_weekend BOOLEAN,
    previous_delay_minutes INTEGER,
    cumulative_delay_minutes INTEGER,
    stops_completed INTEGER,
    stops_remaining INTEGER,
    weather_conditions JSON,
    computed_at TIMESTAMP
);

-- Indexes for efficient querying
CREATE INDEX idx_prediction_results_journey ON prediction_results(journey_id, target_station_code);
CREATE INDEX idx_journey_features_patterns ON journey_features(hour_of_day, day_of_week, is_rush_hour);
```

#### 3.1 Single Train Arrival Prediction

```python
@router.get("/api/v2/predictions/arrival/{train_id}")
async def predict_arrival(
    train_id: str,
    target_station: str,
    date: Optional[str] = None,
    current_position: Optional[str] = None,
    db: AsyncSession = Depends(get_db)
) -> ArrivalPredictionResponse:
    """
    Predict arrival time for a specific train at a target station.
    
    Returns:
    - predicted_arrival: ISO8601 timestamp
    - confidence_interval: {early: timestamp, late: timestamp}
    - delay_probability: 0.0-1.0
    - predicted_delay_minutes: integer
    - confidence_score: 0.0-1.0
    - factors: list of contributing factors
    """
```

#### 3.2 Route Delay Patterns

```python
@router.get("/api/v2/predictions/delay-patterns")
async def get_delay_patterns(
    from_station: str,
    to_station: str,
    time_of_day: Optional[str] = None,
    day_type: Optional[str] = None,  # weekday/weekend/holiday
    db: AsyncSession = Depends(get_db)
) -> DelayPatternResponse:
    """
    Get historical delay patterns for a route.
    
    Returns:
    - average_delay_minutes: float
    - median_delay_minutes: float
    - delay_distribution: {on_time: %, 5min: %, 10min: %, 15min: %, 30min+: %}
    - peak_delay_times: list of time ranges
    - reliability_score: 0.0-1.0
    """
```

#### 4.1 Delay Propagation Model

Account for how delays cascade through a journey:

```python
def propagate_delay(
    current_delay: int,
    current_station: str,
    target_station: str,
    historical_patterns: DelayPatterns
) -> int:
    """
    Model delay propagation based on:
    - Historical recovery rates between stations
    - Schedule padding at intermediate stops
    - Terminal station delay patterns
    - Time of day effects on recovery
    """
```

#### 4.2 Feature Engineering

Extract predictive features from journey data:

```python
async def extract_journey_features(journey: TrainJourney) -> JourneyFeatures:
    return JourneyFeatures(
        hour_of_day=journey.scheduled_departure.hour,
        day_of_week=journey.journey_date.weekday(),
        is_rush_hour=is_rush_hour(journey.scheduled_departure),
        is_holiday=is_holiday(journey.journey_date),
        line_congestion=await calculate_line_congestion(journey),
        historical_performance=await get_train_performance(journey.train_id)
    )
```

#### 4.3 Confidence Scoring

Calculate prediction confidence based on:
- Number of historical samples
- Recency of data
- Variance in historical patterns
- Current conditions vs historical norms

### 5. Machine Learning Pipeline (Future Enhancement)

Prepare for ML model integration:

```python
class DelayPredictionModel:
    """
    Gradient boosting model for delay prediction.
    Features: time patterns, route characteristics, historical performance
    Target: delay_minutes at destination
    """
    
    def train(self, historical_data: DataFrame) -> None:
        # Train on historical journey data
        pass
    
    def predict(self, features: dict) -> DelayPrediction:
        # Return prediction with confidence interval
        pass
```

### 6. Real-Time Updates

Implement WebSocket support for live predictions:

```python
@router.websocket("/ws/predictions/{train_id}")
async def prediction_updates(websocket: WebSocket, train_id: str):
    """Stream updated predictions as train progresses"""
    await websocket.accept()
    while True:
        prediction = await update_prediction(train_id)
        await websocket.send_json(prediction.dict())
        await asyncio.sleep(30)  # Update every 30 seconds
```

#### 7.1 Prediction Service

```swift
class DelayPredictionService {
    static let shared = DelayPredictionService()
    
    func getArrivalPrediction(
        for train: TrainV2,
        at station: String
    ) async throws -> ArrivalPrediction {
        // Call prediction API
        // Cache results
        // Return prediction with UI-friendly format
    }
    
    func getDelayPatterns(
        from: String,
        to: String,
        timeContext: TimeContext? = nil
    ) async throws -> DelayPatterns {
        // Fetch historical patterns
        // Format for display
    }
}
```

#### 7.2 UI Components

```swift
struct ArrivalPredictionView: View {
    let prediction: ArrivalPrediction
    
    var body: some View {
        VStack {
            // Predicted arrival time with confidence interval
            // Delay probability visualization
            // Contributing factors list
        }
    }
}
```

### 8. Monitoring & Feedback Loop

Track prediction accuracy:

```python
@router.post("/api/v2/predictions/feedback")
async def record_prediction_outcome(
    prediction_id: int,
    actual_arrival: datetime,
    db: AsyncSession = Depends(get_db)
):
    """Record actual outcome for accuracy tracking"""
    # Update prediction_results with actual outcome
    # Calculate accuracy metrics
    # Trigger model retraining if needed
```

### Phase 1: Basic Predictions (Week 1-2)
- [ ] Create database schema migrations
- [ ] Implement basic delay prediction using historical averages
- [ ] Add arrival prediction API endpoint
- [ ] Create simple confidence scoring

### Phase 2: Advanced Patterns (Week 3-4)
- [ ] Implement delay propagation model
- [ ] Add time-of-day and day-of-week patterns
- [ ] Create route reliability scoring
- [ ] Add rush hour and holiday factors

### Phase 3: Real-Time Integration (Week 5-6)
- [ ] Implement WebSocket updates
- [ ] Add current position tracking
- [ ] Create iOS prediction service
- [ ] Build UI components for predictions

### Phase 4: Machine Learning (Month 2-3)
- [ ] Prepare training pipeline
- [ ] Train gradient boosting models
- [ ] Implement A/B testing framework
- [ ] Deploy ML predictions alongside rule-based

## Success Metrics

1. **Prediction Accuracy**
   - Arrival time: ±5 minutes for 90% of predictions
   - Delay prediction: ±3 minutes for 85% of predictions
   - Confidence calibration: 90% of high-confidence predictions are accurate

2. **Performance**
   - API response time: <100ms for single predictions
   - Batch predictions: <500ms for route patterns
   - Real-time updates: <2s latency

3. **User Impact**
   - >60% of users view predictions
   - >80% satisfaction with prediction accuracy
   - Reduced missed connections by 25%

## Data Requirements

- Minimum 30 days of historical data for basic predictions
- 90+ days preferred for seasonal patterns
- At least 100 journeys per route for reliable patterns
- Real-time position updates for live predictions

## Error Handling

1. **Insufficient Data**: Fallback to system-wide averages with low confidence
2. **Missing Current Position**: Use last known position or schedule
3. **Extreme Delays**: Cap predictions at reasonable bounds
4. **Service Disruptions**: Flag as unpredictable, provide last known info

## Testing Strategy

1. **Unit Tests**: Algorithm components, feature extraction
2. **Integration Tests**: API endpoints, database queries
3. **Backtesting**: Historical prediction accuracy on past data
4. **A/B Testing**: Compare against simple average delays
5. **Load Testing**: Ensure performance under concurrent requests

## Security Considerations

- No PII in prediction data
- Rate limiting on prediction endpoints
- Cache predictions to prevent abuse
- Audit logging for prediction requests

## Future Enhancements

1. **Weather Integration**: Include weather conditions in predictions
2. **Event Awareness**: Account for major events affecting travel
3. **Cross-Route Predictions**: Predict connections and transfers
4. **Personalized Predictions**: Learn from user's specific journeys
5. **Explanation Engine**: Provide detailed reasoning for predictions

## Dependencies

- Backend: SQLAlchemy, FastAPI, pandas, numpy
- iOS: Swift 5.5+, Combine, SwiftUI
- Infrastructure: PostgreSQL/SQLite, Redis (caching)
- Future: scikit-learn, XGBoost, TensorFlow

## Related Work

- Track Prediction System (see companion issue)
- Pure Data Implementation (PURE_DATA_IMPLEMENTATION_SUMMARY.md)
- Real-time position tracking from journey stops

## Acceptance Criteria

- [ ] API returns predictions with confidence intervals
- [ ] Predictions update in real-time as journey progresses
- [ ] iOS app displays predictions in intuitive format
- [ ] Accuracy metrics meet defined thresholds
- [ ] System handles edge cases gracefully
- [ ] Documentation complete for all components
